### ISSUE TITLE
New Warning for missing translated term blocks submission

### ISSUE THREAD

**Issue by garrett-eclipse** [Mte90/GlotDict#95]
Browser: Chrome Version 59.0.3071.109 (Official Build) (64-bit)
Link: https://translate.wordpress.org/projects/wp-plugins/wordpress-seo/dev/en-ca/default?filters%5Boriginal_id%5D=4792078
Locale used: en_CA

![screen shot 2017-07-13 at 9 40 07 am](https://user-images.githubusercontent.com/8726005/28177387-7d1d52a4-67af-11e7-87ad-0e972cbc738e.png)

Hi @Mte90,

I love the new feature/warning when you're 'Missing term translated using the locale glossary', but I did find it behaves slightly different than other warnings in that it blocks the submission.

With the other warnings like URL changes the string is submitted and marked as Warning, then you can discard the warning and it's left as a translated string.
With the new missing translated term warning it interferes with the submission forcing you to discard and then resubmit.

I hope that makes sense,
Cheers

**Comment by Mte90** [comment]
Uhm so you are saying to remove as warning that block the submission and remove as consequences the discard feature?
This is an experiment for https://github.com/GlotPress/GlotPress-WP/issues/731

**Comment by garrett-eclipse** [comment]
Hi @Mte90,
I definitely love the feature, was just thinking it should function similar to other warnings like when a URL is changed. Upon submitting the status changes to current and the warning is displayed which can be discarded, if you don't discard you can find them under the Warnings filter. 
How your new warning differs is that it blocks the submission keeping the string in as 'untranslated' displays the warning and after discard you have to resubmit. It's that additional resubmit that can be avoided by marking the string as current with warning.
I hope that clarifies a bit. 
Cheers

**Comment by Mte90** [comment]
Uhm I never used the Warnings filter I have to check the behaviour with that.

**Comment by garrett-eclipse** [comment]
Sounds good, looking at the manual it appears Warnings are a status (https://glotpress.blog/the-manual/editing-a-string/).

**Comment by Mte90** [comment]
I need a page with the warnings so I can replicate and test :-)

**Comment by Mte90** [comment]
Ops wrong button

**Comment by garrett-eclipse** [comment]
Hi @Mte90, here's a full page of warnings for you;
https://translate.wordpress.org/projects/wp-plugins/autodescription/dev-readme/en-ca/default?filters%5Bwarnings%5D=yes&filters%5Bstatus%5D=current_or_waiting&sort%5Bby%5D=translation_date_added

*Just discard the warnings when finished testing please as they're all URL warnings that can be ignored as I localized them already

**Comment by Mte90** [comment]
I'm trying to replicate but I don't understand when happen the issue. Can you explain me step by step how happen the problem?

**Comment by garrett-eclipse** [comment]
Sure thing @Mte90 

First I'll go through the process of an existing GlotPress warning such as the URL changed warning to illustrate the difference in process.
1. Open untranslated row containing a wordpress.org URL
![screen shot 2017-07-21 at 1 30 00 pm](https://user-images.githubusercontent.com/8726005/28481219-e49f549e-6e18-11e7-8d4c-60ac37794606.png)
2. Copy original into textarea
![screen shot 2017-07-21 at 1 30 06 pm](https://user-images.githubusercontent.com/8726005/28481222-e81fd878-6e18-11e7-9ba1-405509e28823.png)
3. Edit textarea so wordpress.org url is localized (ie. en-ca.wordpress.org)
![screen shot 2017-07-21 at 1 30 14 pm](https://user-images.githubusercontent.com/8726005/28481232-ef45ddb4-6e18-11e7-9760-ee6bb8d0bc12.png)
4. Click Add Translation button (*The row will stay open with the warning displayed but it's status is now current)
![screen shot 2017-07-21 at 1 30 21 pm](https://user-images.githubusercontent.com/8726005/28481248-05fb19f2-6e19-11e7-9872-6500c2a6228b.png)
5. Close to row (You'll see the red line indicating a warning and the green indicating it's a current translated string)
![screen shot 2017-07-21 at 1 30 31 pm](https://user-images.githubusercontent.com/8726005/28481275-1d25fdf4-6e19-11e7-941a-1486277438a7.png)
At this point you could move on and it would be seen as translated, but would be findable within the warning filter. If you discard the warning it'll be seen as a translated string.
![screen shot 2017-07-21 at 1 30 46 pm](https://user-images.githubusercontent.com/8726005/28481303-3a27c0c2-6e19-11e7-9246-604eae1d0ae3.png)


Now I'll go through the process that occurs with the new missing translation term warning and how it blocks the add translate and forces you to resubmit once you've discarded the warning.
1. Open untranslated row containing a wordpress.org URL
![screen shot 2017-07-21 at 1 35 29 pm](https://user-images.githubusercontent.com/8726005/28481400-9c500106-6e19-11e7-9981-0c911c7c68ca.png)
2. Copy original into textarea
![screen shot 2017-07-21 at 1 35 34 pm](https://user-images.githubusercontent.com/8726005/28481404-a066a682-6e19-11e7-8402-079945394d34.png)
Leave the string as is without replacing the translation term (so in this case not localizing the URL)
4. Click Add Translation button (*The row will stay open with the warning displayed but _**it's status is left as untranslated**_)
![screen shot 2017-07-21 at 1 35 41 pm](https://user-images.githubusercontent.com/8726005/28481457-e1243644-6e19-11e7-9798-1f8e583bc581.png)
If you close the row the string will be left as untranslated, if you reload the textarea is cleared losing your translation. You have to discard the warning and submit a second time for it to accept your translation.


Hopefully that clarifies, the main problem is it blocks you from submitting your translation until you discard or handle the warning. As such if you reload you lose your translation. In many cases people's workflow is to continue past warnings and come back using the warning filter to handle all warnings together, but with this new warning when you come back your translation is lost and you have to retranslate.

I hope that makes sense, mainly the warning shouldn't block the submission it should keep the translation submission and just add the warning to it.

Let me know if anything is unclear still and I can setup a screencast.

Cheers

**Comment by Mte90** [comment]
Got it finally!
The initial problem is on GlotPress because the warning of glotdict has the class in the div so is not GD that create the problem :-/

**Comment by garrett-eclipse** [comment]
Thanks @Mte90 glad we got it sorted, sorry for all the confusion

**Comment by fxbenard** [comment]
Great explanation, exactly what happens

**Comment by garrett-eclipse** [comment]
@Mte90 taking another look at this I think the issue myself and @fxbenard are having is that the validation for glossary term is done on.click and due to the validation failure it halts the translation submission.

Diving into code there's an onFirst call [here](https://github.com/Mte90/GlotDict/blob/b34ab4b0866444e303700af0cd43d317e706af8c/js/glotdict.js#L42);
`jQuery($gp.editor.table).onFirst('click', 'button.ok', gd_validate_visible);`

This executes the [gd_validate_visible](https://github.com/Mte90/GlotDict/blob/b34ab4b0866444e303700af0cd43d317e706af8c/js/glotdict-validation.js#L116) function which in turn calls [gd_validate](https://github.com/Mte90/GlotDict/blob/b34ab4b0866444e303700af0cd43d317e706af8c/js/glotdict-validation.js#L68). Within gd_validate you increment howmany due to the [gd_search_glossary_on_translation](https://github.com/Mte90/GlotDict/blob/b34ab4b0866444e303700af0cd43d317e706af8c/js/glotdict-validation.js#L27). Due to howmany being positive due to the warning propagation is halted stoping submission through the [gd_stoppropagation](https://github.com/Mte90/GlotDict/blob/bab1574bd5898c330552611737d79731f563be65/js/glotdict-functions.js#L231) function.

TLDR; The warning stops translation submission.

So as to match Glotpress' behaviour for warnings I'd suggest the GlotDict validation be done after submission. 

And maybe because some warnings should block submission such as the empty textarea maybe split validation into two so errors block submission and warnings allow submission but still flag the issue.

Hope that makes sense

**Comment by Mte90** [comment]
There was an issue on `gd_validate` that was blocking the execution, now with this little fix is everything :-)

**Comment by garrett-eclipse** [comment]
Thanks @Mte90 I'll give it a shot once you push your next version. Cheers

**Comment by Mte90** [comment]
I released today a new version :-)

**Comment by fxbenard** [comment]
Same here, let’s wait for this update, cheers

**Comment by garrett-eclipse** [comment]
Hi @Mte90 sorry for the delay got around to testing this and I'm not being blocked from submission so that's good thank you. But now there's no warning whatsoever for the missing glossary translation.

**Comment by Mte90** [comment]
Can you open a new ticket and explain better the issue?

**Comment by garrett-eclipse** [comment]
Nvmd, sorry @Mte90 I forgot I had disabled that setting by checking it off. Unchecking 'No glossary term missing warning on translation' the warnings have returned and they're still blocking submission which is the original issue myself and @fxbenard raised through this ticket and #117.
The main issue both of us were having is that warning blocks string submission interrupting translation flow, so if we can move the warning to post-submission so it doesn't block the submit but still is flagged as a warning I believe that would resolve mine and @fxbenard issue with this particular warning.
Thank you
